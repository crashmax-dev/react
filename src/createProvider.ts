import { createContext } from 'react'
import { useContextProvider } from './useContextProvider.js'

/**
 * Creates a context provider, hook, and consumer
 * @param displayName The name of the context. Used for error messages
 * @param defaultValue The default value of the context
 * @returns A tuple of the hook, provider, and consumer
 */
export function createProvider<T>(
  displayName: string,
  defaultValue: T | null = null
): readonly [
  hook: () => T,
  provider: React.Provider<T | null>,
  consumer: React.Consumer<T | null>
] {
  const context = createContext<T | null>(defaultValue)
  const hook = () => useContextProvider(displayName, context)
  const provider = context.Provider
  const consumer = context.Consumer

  return [
    hook,
    provider,
    consumer
  ]
}
