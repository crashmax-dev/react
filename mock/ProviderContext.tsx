import React, { useState } from 'react'
import { createProvider } from '../src/createProvider.js'

interface Counter {
  count: number
  setCount: React.Dispatch<React.SetStateAction<number>>
}

const [useCounter, CounterProvider] = createProvider<Counter>('Counter')

function Counter() {
  const { count, setCount } = useCounter()

  return (
    <div>
      <h1 aria-label="count">{count}</h1>
      <button
        aria-label="button"
        onClick={() => setCount(count + 1)}
      >
        Increment
      </button>
    </div>
  )
}

export function App() {
  const [count, setCount] = useState(0)

  return (
    <CounterProvider value={{ count, setCount }}>
      <Counter />
    </CounterProvider>
  )
}

export function AppWithError() {
  const counter = useCounter() // ERROR
  const [count, setCount] = useState(0)

  return (
    <CounterProvider value={{ count, setCount }}>
      <Counter />
    </CounterProvider>
  )
}
