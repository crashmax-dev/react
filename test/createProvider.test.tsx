import { App, AppWithError } from '../mock/ProviderContext.js'
import { createProvider } from '../src/createProvider.js'
import { fireEvent, render, screen } from './utils.js'

describe('createProvider', () => {
  it('should be defined', () => {
    expect(createProvider).toBeDefined()
  })

  it('should return an array', () => {
    const [
      useFoo,
      FooProvider,
      FooConsumer
    ] = createProvider('Foo')

    expect(useFoo).toBeDefined()
    expect(FooProvider).toBeDefined()
    expect(FooConsumer).toBeDefined()
  })

  it('should render the app', () => {
    render(<App />)
    const count = screen.getByLabelText('count')
    const button = screen.getByLabelText('button')

    expect(count).toHaveTextContent('0')
    expect(button).toHaveTextContent('Increment')

    fireEvent.click(button)
    expect(count).toHaveTextContent('1')
  })

  it('should throw an error if used outside of a provider', () => {
    expect(() => render(<AppWithError />)).toThrowErrorMatchingSnapshot()
  })
})
