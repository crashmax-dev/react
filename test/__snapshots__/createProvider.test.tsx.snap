// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createProvider > should throw an error if used outside of a provider 1`] = `"useCounter must be used within a CounterProvider"`;
